File: zero-gateway/pkg/filters/metrics.go
package filters

import (
	"sync"
	"time"
)

// FilterMetrics 过滤器性能指标
type FilterMetrics struct {
	mu sync.RWMutex

	// 执行统计
	TotalExecutions   int64            `json:"total_executions"`
	SuccessfulExecutions int64         `json:"successful_executions"`
	FailedExecutions  int64            `json:"failed_executions"`
	
	// 性能统计
	TotalDuration     time.Duration   `json:"total_duration"`
	MinDuration       time.Duration   `json:"min_duration"`
	MaxDuration       time.Duration   `json:"max_duration"`
	AvgDuration       time.Duration   `json:"avg_duration"`
	
	// 按过滤器名称的统计
	FilterStats       map[string]*FilterStats `json:"filter_stats"`
}

// FilterStats 单个过滤器的统计信息
type FilterStats struct {
	Name              string          `json:"name"`
	TotalExecutions   int64           `json:"total_executions"`
	SuccessfulExecutions int64        `json:"successful_executions"`
	FailedExecutions  int64           `json:"failed_executions"`
	TotalDuration     time.Duration   `json:"total_duration"`
	MinDuration       time.Duration   `json:"min_duration"`
	MaxDuration       time.Duration   `json:"max_duration"`
	AvgDuration       time.Duration   `json:"avg_duration"`
}

// NewFilterMetrics 创建过滤器指标收集器
func NewFilterMetrics() *FilterMetrics {
	return &FilterMetrics{
		FilterStats: make(map[string]*FilterStats),
		MinDuration: time.Hour, // 初始化为一个很大的值
	}
}

// RecordExecution 记录过滤器执行
func (fm *FilterMetrics) RecordExecution(filterName string, success bool, duration time.Duration) {
	fm.mu.Lock()
	defer fm.mu.Unlock()

	// 更新总体统计
	fm.TotalExecutions++
	if success {
		fm.SuccessfulExecutions++
	} else {
		fm.FailedExecutions++
	}

	fm.TotalDuration += duration
	if duration < fm.MinDuration {
		fm.MinDuration = duration
	}
	if duration > fm.MaxDuration {
		fm.MaxDuration = duration
	}
	fm.AvgDuration = fm.TotalDuration / time.Duration(fm.TotalExecutions)

	// 更新单个过滤器统计
	stats, exists := fm.FilterStats[filterName]
	if !exists {
		stats = &FilterStats{
			Name:            filterName,
			MinDuration:     time.Hour,
		}
		fm.FilterStats[filterName] = stats
	}

	stats.TotalExecutions++
	if success {
		stats.SuccessfulExecutions++
	} else {
		stats.FailedExecutions++
	}

	stats.TotalDuration += duration
	if duration < stats.MinDuration {
		stats.MinDuration = duration
	}
	if duration > stats.MaxDuration {
		stats.MaxDuration = duration
	}
	stats.AvgDuration = stats.TotalDuration / time.Duration(stats.TotalExecutions)
}

// GetMetrics 获取指标快照
func (fm *FilterMetrics) GetMetrics() FilterMetrics {
	fm.mu.RLock()
	defer fm.mu.RUnlock()

	// 创建副本
	snapshot := FilterMetrics{
		TotalExecutions:      fm.TotalExecutions,
		SuccessfulExecutions: fm.SuccessfulExecutions,
		FailedExecutions:     fm.FailedExecutions,
		TotalDuration:        fm.TotalDuration,
		MinDuration:          fm.MinDuration,
		MaxDuration:          fm.MaxDuration,
		AvgDuration:          fm.AvgDuration,
		FilterStats:          make(map[string]*FilterStats),
	}

	// 复制过滤器统计
	for name, stats := range fm.FilterStats {
		snapshot.FilterStats[name] = &FilterStats{
			Name:                stats.Name,
			TotalExecutions:    stats.TotalExecutions,
			SuccessfulExecutions: stats.SuccessfulExecutions,
			FailedExecutions:     stats.FailedExecutions,
			TotalDuration:        stats.TotalDuration,
			MinDuration:          stats.MinDuration,
			MaxDuration:          stats.MaxDuration,
			AvgDuration:          stats.AvgDuration,
		}
	}

	return snapshot
}

// Reset 重置指标
func (fm *FilterMetrics) Reset() {
	fm.mu.Lock()
	defer fm.mu.Unlock()

	fm.TotalExecutions = 0
	fm.SuccessfulExecutions = 0
	fm.FailedExecutions = 0
	fm.TotalDuration = 0
	fm.MinDuration = time.Hour
	fm.MaxDuration = 0
	fm.AvgDuration = 0
	fm.FilterStats = make(map[string]*FilterStats)
}

File: zero-agent/filters/metrics.py
"""
过滤器性能指标收集
"""

import time
from typing import Dict, Optional
from dataclasses import dataclass, field
from threading import Lock


@dataclass
class FilterStats:
    """单个过滤器的统计信息"""
    name: str
    total_executions: int = 0
    successful_executions: int = 0
    failed_executions: int = 0
    total_duration_ns: int = 0
    min_duration_ns: int = 0
    max_duration_ns: int = 0

    @property
    def avg_duration_ns(self) -> int:
        """平均执行时间（纳秒）"""
        if self.total_executions == 0:
            return 0
        return self.total_duration_ns // self.total_executions

    @property
    def avg_duration_ms(self) -> float:
        """平均执行时间（毫秒）"""
        return self.avg_duration_ns / 1_000_000


@dataclass
class FilterMetrics:
    """过滤器性能指标"""
    total_executions: int = 0
    successful_executions: int = 0
    failed_executions: int = 0
    total_duration_ns: int = 0
    min_duration_ns: int = 0
    max_duration_ns: int = 0
    filter_stats: Dict[str, FilterStats] = field(default_factory=dict)
    _lock: Lock = field(default_factory=Lock)

    @property
    def avg_duration_ns(self) -> int:
        """平均执行时间（纳秒）"""
        if self.total_executions == 0:
            return 0
        return self.total_duration_ns // self.total_executions

    @property
    def avg_duration_ms(self) -> float:
        """平均执行时间（毫秒）"""
        return self.avg_duration_ns / 1_000_000

    def record_execution(self, filter_name: str, success: bool, duration_ns: int):
        """记录过滤器执行"""
        with self._lock:
            # 更新总体统计
            self.total_executions += 1
            if success:
                self.successful_executions += 1
            else:
                self.failed_executions += 1

            self.total_duration_ns += duration_ns
            if self.min_duration_ns == 0 or duration_ns < self.min_duration_ns:
                self.min_duration_ns = duration_ns
            if duration_ns > self.max_duration_ns:
                self.max_duration_ns = duration_ns

            # 更新单个过滤器统计
            if filter_name not in self.filter_stats:
                self.filter_stats[filter_name] = FilterStats(name=filter_name)

            stats = self.filter_stats[filter_name]
            stats.total_executions += 1
            if success:
                stats.successful_executions += 1
            else:
                stats.failed_executions += 1

            stats.total_duration_ns += duration_ns
            if stats.min_duration_ns == 0 or duration_ns < stats.min_duration_ns:
                stats.min_duration_ns = duration_ns
            if duration_ns > stats.max_duration_ns:
                stats.max_duration_ns = duration_ns

    def get_metrics(self) -> Dict:
        """获取指标快照"""
        with self._lock:
            return {
                "total_executions": self.total_executions,
                "successful_executions": self.successful_executions,
                "failed_executions": self.failed_executions,
                "avg_duration_ms": self.avg_duration_ms,
                "min_duration_ms": self.min_duration_ns / 1_000_000,
                "max_duration_ms": self.max_duration_ns / 1_000_000,
                "filter_stats": {
                    name: {
                        "name": stats.name,
                        "total_executions": stats.total_executions,
                        "successful_executions": stats.successful_executions,
                        "failed_executions": stats.failed_executions,
                        "avg_duration_ms": stats.avg_duration_ms,
                        "min_duration_ms": stats.min_duration_ns / 1_000_000,
                        "max_duration_ms": stats.max_duration_ns / 1_000_000,
                    }
                    for name, stats in self.filter_stats.items()
                }
            }

    def reset(self):
        """重置指标"""
        with self._lock:
            self.total_executions = 0
            self.successful_executions = 0
            self.failed_executions = 0
            self.total_duration_ns = 0
            self.min_duration_ns = 0
            self.max_duration_ns = 0
            self.filter_stats.clear()
